src/processor/operator/persistent/node_batch_insert.cpp

```diff
@@ -207,19 +219,17 @@ void NodeBatchInsert::appendIncompleteNodeGroup(transaction::Transaction* transa
 
 void NodeBatchInsert::finalize(ExecutionContext* context) {
     const auto nodeSharedState = ku_dynamic_cast<NodeBatchInsertSharedState*>(sharedState.get());
+    auto errorHandler = createErrorHandler(context);
     if (nodeSharedState->sharedNodeGroup) {
         while (nodeSharedState->sharedNodeGroup->getNumRows() > 0) {
             writeAndResetNodeGroup(context->clientContext->getTx(),
                 nodeSharedState->sharedNodeGroup, nodeSharedState->globalIndexBuilder,
-                context->clientContext->getMemoryManager());
+                context->clientContext->getMemoryManager(), errorHandler);
         }
     }
     if (nodeSharedState->globalIndexBuilder) {
-        auto* localNodeState = ku_dynamic_cast<NodeBatchInsertLocalState*>(localState.get());
-        KU_ASSERT(localNodeState->errorHandler.has_value());
-        nodeSharedState->globalIndexBuilder->finalize(context,
-            localNodeState->errorHandler.value());
-        localNodeState->errorHandler->flushStoredErrors();
+        nodeSharedState->globalIndexBuilder->finalize(context, errorHandler);
```

Comment:
Just to help guard against misuse can you add:
```suggestion
        KU_ASSERT(!localNodeState->errorHandler.has_value())
```